bpf/lib/nat.h

```diff
@@ -852,18 +852,19 @@ __snat_v4_nat(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple, fraginfo_t fr
 	      int l4_off, bool update_tuple, const struct ipv4_nat_target *target,
 	      __u16 port_off, struct trace_ctx *trace, __s8 *ext_err)
 {
-	struct ipv4_nat_entry *state, tmp;
+	struct ipv4_nat_entry *state = NULL, tmp;
 	int ret;
 
 	ret = snat_v4_nat_handle_mapping(ctx, tuple, fraginfo, &state, &tmp,
 					 l4_off, target, trace, ext_err);
 	if (ret < 0)
 		return ret;
 
-	ret = snat_v4_rewrite_headers(ctx, tuple->nexthdr, ETH_HLEN,
-				      ipfrag_has_l4_header(fraginfo), l4_off,
-				      tuple->saddr, state->to_saddr, IPV4_SADDR_OFF,
-				      tuple->sport, state->to_sport, port_off);
+	if (state)
+		ret = snat_v4_rewrite_headers(ctx, tuple->nexthdr, ETH_HLEN,
+					      ipfrag_has_l4_header(fraginfo), l4_off,
+					      tuple->saddr, state->to_saddr, IPV4_SADDR_OFF,
+					      tuple->sport, state->to_sport, port_off);
 
 	if (update_tuple) {
 		tuple->saddr = state->to_saddr;
```

Comment:
Curious that we need to guard the initial access to `state`, but not these additional accesses below ...

Reply:
1. The kernel version I used is similar to this one:
https://gitee.com/OpenCloudOS/OpenCloudOS-Kernel/tree/linux-5.4%2Flts%2F5.4.241-30.0017/
2. I looked at the commit history of the kernel/bpf/verifier.c, this file is indeed a bit old. Therefore, the most likely reason is that the bpf verifier rules are relatively old due to the lack of some new commits.
3. By the way, is the initialization method of state here also a bit unstandardized?
From the snat_v4_nat_handle_mapping function alone, there is no comment requiring that state must not be NULL.